core/src/main/java/org/apache/iceberg/ManifestExpirationManager.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.io.UncheckedIOException;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import org.apache.commons.compress.utils.Sets;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ManifestExpirationManager implements Serializable {
+  private static final Logger LOG = LoggerFactory.getLogger(ManifestExpirationManager.class);
+
+  private final Set<Long> validSnapshotIds;
+  private final Map<Integer, PartitionSpec> specLookup;
+  private final FileIO io;
+
+  private ManifestExpirationManager(Set<Long> validSnapshotIds, Map<Integer, PartitionSpec> specLookup, FileIO io) {
+    this.validSnapshotIds = validSnapshotIds;
+    this.specLookup = specLookup;
+    this.io = io;
+  }
+
+  /**
+   * Creates a Manifest Scanner for analyzing manifests for files which are no longer needed
+   * after expiration.
+   * @param current metadata for the table being expired
+   * @param io FileIO for the table, for Serializable usecases make sure this is also serializable for the framework
```

Comment:
I don't follow the javadoc on the param completely.
> for Serializable usecases make sure this is also serializable for the framework

Is that meant as a TODO?

Reply:
Unfortunately we can't control the underlying implementation of IO here and some instances will require a specfic manipulation to make sure they serialize correctly.

I added this specifically (instead of just allowing passing in ops) because for Spark we use SparkUtil.serializableIO to check whether the IO is of the HadoopIO class and make it serializable if so. We can't do that here because it would require dependencies from other modules and also be more or less specific to Spark. I think our only real option is to leave it up to the caller to make sure that portion of the class is correct for their application. :/